NVDAObjects.IAccessible: output alerts in braille

josephsl opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft. See https://github.com/nvaccess/nvda/wiki/Contributing
-->

### Link to issue number:
Fixes #14562 

### Summary of the issue:
Alert event text is not shown on braille displays when the event occurs such as when downloading files via Google Chrome.

### Description of user facing changes
NVDA will present alert messages in braille.

### Description of development approach
Add braille output (braille.handler.message) to NVDAObjects.IAccessible.IAccessible.event_alert, borrowing heavily from notification behavior.

### Testing strategy:
Manual testing:

1. Visit websites such as community add-ons website (addons.nvda-project.org) where files can be downloaded.
2. Download a file while using a braille display.

Before the PR: speech output only when alert event is fired.
After the PR: speech and braille outpu ocurs.

### Known issues with pull request:
While Chrom's file download alert provides a good test case for testing braille output for alert event,s it might be possible that apps and web elements may not fire alert events, or if they do, presented in a way that NVDA doesn't know about (alert events are handled if the control role is alert).

### Change log entries:
Bug fixes:

In web browsers such as Chrome and Firefox, alerts such as file downloads via Chrome are shown in braille in addition to being spoken. (#14562)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

josephsl committed:
NVDAObjects.IAccessible: report alerts in braile. Re #14562.

When processing MSAA alert event, only speech output was invoked whereas other implementations such as notification behavior class also brailled alerts and notifications. Therefore add braille output to alert event handler, both for the root alert and its children, borrowing heavily from notification behavior.

josephsl committed:
NVDAObjects.IAccessible: update copyright year

josephsl committed:
NVDAObjects.IAccessible: add type information for alert event. Re #14562

josephsl commented:
Hi,

I think this might be a good opportunity to introduce a generic alert event to base NVDA object or the notification behavior (perhaps in 2023.x or 2024.1) unless alert event is specific to IAccessible and/or UIA. Among other things, the default implementation would allow vision enhancement providers to offer a way to look at where the alert is coming from.

Thanks.

seanbudd commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -1702,9 +1702,12 @@ def event_alert(self):
 		if self in api.getFocusAncestors():
 			return
 		speech.speakObject(self, reason=controlTypes.OutputReason.FOCUS, priority=speech.Spri.NOW)
+		# Ideally, we wouldn't use getPropertiesBraille directly.
```
what would be better? why not do that? it would be good to include this in the comment.

seanbudd replied:
it appears that using this directly is common practice, I think this comment can be droppped

seanbudd commented:
>I think this might be a good opportunity to introduce a generic alert event to base NVDA object or the notification behavior (perhaps in 2023.x or 2024.1) unless alert event is specific to IAccessible and/or UIA. Among other things, the default implementation would allow vision enhancement providers to offer a way to look at where the alert is coming from.

Would you be interested in doing this in this PR?

josephsl commented:
Hi, not in this PR as this addresses a specific alert case (UIA is handled through system alert event); I think I’m seeing a mentoring opportunity for someone. Thanks for the suggestion (will think about it in northern summer once spring classes are done).

josephsl committed:
Merge branch 'master' into i14562IAccessibleAlertBrailleOutput

josephsl committed:
NVDAObjects.IAccessible: remove braille message comment.

Suggestion from Sean Budd (NV Access): since braille.handler.message comment is found in other places and the actual method is used elsewhere, remove the message comment.

seanbudd committed:
update changes

seanbudd committed:
fixup change
